### ISSUE TITLE
Iterate over infinite reccurence.

### ISSUE THREAD

**Issue by akhramov** [fmeringdal/rust_rrule#8]
Let's suppose I have an infinite RRule such as 

```
DTSTART:2022-02-01T10:00:00Z
RRULE:FREQ=HOURLY
```

Since rrule.all returns Vec, it would likely try to create an infinite vec.
https://github.com/fmeringdal/rust_rrule/blob/main/src/rrule.rs#L20-L31

Is there a good way to deal with infinite reccurences?

I imagine having an Iterator instead of Vec...

**Comment by fmeringdal** [comment]
Unfortunately, there is currently no good way to handle an infinite recurrence rule as this crate does not support iterators. That is because this crate is largely a port of the javascript rrule library and javascript has a very different way of iterating that rust have.
That said, this is something I have had in the backlog for some time now, and hopefully will implement very soon (maybe this weekend)!

**Comment by akhramov** [comment]
Ah, thanks for the swift response. The library saved me lots of time already and I really like it.

FWIW, I think you have everything you need already. I implemented an iterator using functions exposed by the library.

So! Probably there's no need for a big refactoring.

```rust
struct RruleIterator {
    rule: RRule,
    prev: DateTime<Tz>,
}

impl RruleIterator {
    fn new(rule: RRule) -> Self {
        Self {
            rule,
            prev: Tz::UTC.timestamp_nanos(0),
        }
    }
}

impl Iterator for RruleIterator {
    type Item = DateTime<Tz>;

    fn next(&mut self) -> Option<Self::Item> {
        let next = self.rule.after(self.prev, true);

        if let Some(value) = next {
            self.prev = value;
        }

        next
    }
}
```

I am not sure if I understood the second parameter of `rule.after` correctly, though

**Comment by fmeringdal** [comment]
I am very glad to hear you liked this library @akhramov, I have put a lot of effort into it  :)

Your iterator implementation looks correct, except I think it should be `self.rule.after(self.prev, false)`. That solution is also quite slow since every call to time you call next it will regenerate all the different masks (library uses a lot of masks internally to filter out invalid dates) which is expensive.

**Comment by fmeringdal** [comment]
Happy news! 
After a big refactoring, iterators are finally supported for both the `RRule` and `RRuleSet` types in version 0.5.0.
Take a look at the documentation for some examples of using them: [rrule docs](https://docs.rs/rrule/0.5.0/rrule/).

Please let me know if you discover any bugs or want some other improvements for this crate :)

**Comment by akhramov** [comment]
Wow, thank you!

I am testing the new functionality and it works really great on most of the cases.

Alas, I found a buggy case (a panic), here's the repro

```rust
let mut rrule = "DTSTART:20220201T100000Z\nRRULE:FREQ=DAILY"
    .parse::<RRule>()?;
rrule.clone().into_iter().nth(1500000);
```

The panic

```
thread 'test_name' panicked at 'index out of bounds: the len is 365 but the index is 365', /usr/home/akhramov/.cargo/registry/src/github.com-1ecc6299db9ec823/rrule-0.5.0/src/iter/iterinfo.rs:187:9
```

I was able to reproduce the problem on daily rrules, only.

The first problematic (panicking) event is somewhere at years 2200-2201. For following rrule it fails on day 364:

`DTSTART:22010101T100000Z\nRRULE:FREQ=DAILY`

Please let me know if I can help you somehow, thanks!

Happily, the bug doesn't affect my current needs :)

**Comment by fmeringdal** [comment]
Oh, nice catch!
I have fixed the problem in version 0.5.2. The new maximum limit is year 9999, which will just stop the iteration when passed and not panic.
